vm/cuttlefish/cuttlefish.go

```diff
@@ -0,0 +1,182 @@
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+// Package cuttlefish allows to use Cuttlefish Android emulators hosted on Google Compute Engine
+// (GCE) virtual machines as VMs. It is assumed that syz-manager also runs on GCE as VMs are
+// created in the current project/zone.
+//
+// See https://cloud.google.com/compute/docs for details.
+// In particular, how to build GCE-compatible images:
+// https://cloud.google.com/compute/docs/tutorials/building-images
+package cuttlefish
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/google/syzkaller/pkg/config"
+	"github.com/google/syzkaller/pkg/gce"
+	"github.com/google/syzkaller/pkg/log"
+	"github.com/google/syzkaller/pkg/report"
+	"github.com/google/syzkaller/vm/vmimpl"
+
+	gcevm "github.com/google/syzkaller/vm/gce"
+)
+
+func init() {
+	vmimpl.Register("cuttlefish", ctor, true)
+}
+
+type Config struct {
+	Count       int    `json:"count"`        // number of VMs to use
+	MachineType string `json:"machine_type"` // GCE machine type (e.g. "n1-standard-4")
+	GCEImage    string `json:"gce_image"`    // pre-created GCE image to use
+	Preemptible bool   `json:"preemptible"`  // use preemptible VMs if available (defaults to true)
+}
+
+type Pool struct {
+	cfg     *Config
+	env     *vmimpl.Env
+	gcePool *gcevm.Pool
+}
+
+type instance struct {
+	debug   bool
+	gceInst vmimpl.Instance
+}
+
+func ctor(env *vmimpl.Env) (vmimpl.Pool, error) {
+	log.Logf(0, "cuttlefish.ctor()")
+	if env.Name == "" {
+		return nil, fmt.Errorf("config param name is empty (required for GCE)")
+	}
+	cfg := &Config{
+		Count:       1,
+		Preemptible: true,
+	}
+	if err := config.LoadData(env.Config, cfg); err != nil {
+		return nil, fmt.Errorf("failed to parse gce vm config: %v", err)
+	}
+	if cfg.Count < 1 || cfg.Count > 1000 {
+		return nil, fmt.Errorf("invalid config param count: %v, want [1, 1000]", cfg.Count)
+	}
+	if env.Debug && cfg.Count > 1 {
+		log.Logf(0, "limiting number of VMs from %v to 1 in debug mode", cfg.Count)
+		cfg.Count = 1
+	}
+	if cfg.MachineType == "" {
+		return nil, fmt.Errorf("machine_type parameter is empty")
+	}
+	if cfg.GCEImage == "" {
+		return nil, fmt.Errorf("gce_image parameter is empty")
+	}
+
+	ctx, err := gce.NewContext()
+	if err != nil {
+		return nil, fmt.Errorf("failed to init gce: %v", err)
+	}
+	log.Logf(0, "GCE initialized: running on %v, internal IP %v, project %v, zone %v, net %v/%v",
+		ctx.Instance, ctx.InternalIP, ctx.ProjectID, ctx.ZoneID, ctx.Network, ctx.Subnetwork)
+
+	pool := &Pool{
+		cfg: cfg,
+		env: env,
+		// This nested gcevm.Pool object will let us re-use the existing Create() function.
+		gcePool: &gcevm.Pool{
+			Env: env,
+			Cfg: &gcevm.Config{
+				Count:       cfg.Count,
+				MachineType: cfg.MachineType,
+				GCEImage:    cfg.GCEImage,
+				Preemptible: cfg.Preemptible,
+			},
+			GCE: ctx,
+		},
+	}
+
+	return pool, nil
+}
+
+func (pool *Pool) Count() int {
+	log.Logf(0, "cuttlefish.pool.Count()")
```

Comment:
Please use at least log level 1. Level 0 will pollute manager output always.
For debugging you can start syz-manager with -vv=10

Reply:
Done!